experimental-utils v3.0 requires TS 3.8 compiler and ES2015 target

I just upgraded to typescript-eslint 3.0 in a TypeScript project which uses `@typescript-eslint/experimental-utils` and got a slightly surprising [compile error](https://app.circleci.com/pipelines/github/palantir/blueprint/166/workflows/6a310246-8687-4704-92a2-ab9d049dd428/jobs/32657/steps). The usage of ES private field syntax introduced in #2023 requires downstream TypeScript consumers to target ES2015 / ES6 or newer ([TS docs](https://devblogs.microsoft.com/typescript/announcing-typescript-3-8-beta/#ecmascript-private-fields)).

```
../../node_modules/@typescript-eslint/experimental-utils/dist/eslint-utils/RuleTester.d.ts:7:5 - error TS18028: Private identifiers are only available when targeting ECMAScript 2015 and higher.
7     #private;
      ~~~~~~~~
```

This isn't a big problem for me, since I can update my `"target"` compiler option, but I just wanted to note this slight breaking change.

**Versions**

| package                                 | version |
| --------------------------------------- | ------- |
| `@typescript-eslint/experimental-utils` | `3.0.1` |
| `TypeScript`                            | `3.9.3` |
| `node`                                  | `12.16.2` |

--- a/.gitignore
+++ b/.gitignore
@@ -64,5 +64,6 @@ jspm_packages/
 .DS_Store
 .idea
 dist
+_ts3.4
 *.tsbuildinfo
 .watchmanconfig

--- a/package.json
+++ b/package.json
@@ -49,7 +49,18 @@
     "extends": [
       "@commitlint/config-conventional",
       "@commitlint/config-lerna-scopes"
-    ]
+    ],
+    "rules": {
+      "body-max-length": [
+        0
+      ],
+      "footer-max-length": [
+        0
+      ],
+      "header-max-length": [
+        0
+      ]
+    }
   },
   "engines": {
     "node": "^10.12.0 || >=12.0.0"
@@ -64,6 +75,7 @@
     "all-contributors-cli": "^6.14.2",
     "cspell": "^4.0.61",
     "cz-conventional-changelog": "^3.2.0",
+    "downlevel-dts": "^0.4.0",
     "eslint": "^7.0.0",
     "eslint-plugin-eslint-comments": "^3.1.2",
     "eslint-plugin-eslint-plugin": "^2.2.1",
@@ -76,6 +88,7 @@
     "lint-staged": "^10.2.2",
     "markdownlint-cli": "^0.23.0",
     "prettier": "^2.0.5",
+    "rimraf": "^3.0.2",
     "ts-jest": "^25.5.1",
     "ts-node": "^8.10.1",
     "tslint": "^6.1.2",

--- a/packages/eslint-plugin-internal/package.json
+++ b/packages/eslint-plugin-internal/package.json
@@ -6,6 +6,7 @@
   "scripts": {
     "build": "tsc -b tsconfig.build.json",
     "clean": "tsc -b tsconfig.build.json --clean",
+    "postclean": "rimraf dist",
     "format": "prettier --write \"./**/*.{ts,js,json,md}\" --ignore-path ../../.prettierignore",
     "lint": "eslint . --ext .js,.ts --ignore-path='../../.eslintignore'",
     "test": "jest --coverage",

--- a/packages/eslint-plugin-tslint/package.json
+++ b/packages/eslint-plugin-tslint/package.json
@@ -25,6 +25,7 @@
   "scripts": {
     "build": "tsc -b tsconfig.build.json",
     "clean": "tsc -b tsconfig.build.json --clean",
+    "postclean": "rimraf dist",
     "format": "prettier --write \"./**/*.{ts,js,json,md}\" --ignore-path ../../.prettierignore",
     "lint": "eslint . --ext .js,.ts --ignore-path='../../.eslintignore'",
     "test": "jest --coverage",

--- a/packages/eslint-plugin/package.json
+++ b/packages/eslint-plugin/package.json
@@ -33,6 +33,7 @@
     "check:docs": "jest tests/docs.test.ts --runTestsByPath --silent --runInBand",
     "check:configs": "jest tests/configs.test.ts --runTestsByPath --silent --runInBand",
     "clean": "tsc -b tsconfig.build.json --clean",
+    "postclean": "rimraf dist",
     "format": "prettier --write \"./**/*.{ts,js,json,md}\" --ignore-path ../../.prettierignore",
     "generate:configs": "../../node_modules/.bin/ts-node --files --transpile-only tools/generate-configs.ts",
     "generate:rules-lists": "../../node_modules/.bin/ts-node --files --transpile-only tools/generate-rules-lists.ts",

--- a/packages/experimental-utils/package.json
+++ b/packages/experimental-utils/package.json
@@ -12,6 +12,7 @@
   },
   "files": [
     "dist",
+    "_ts3.4",
     "package.json",
     "README.md",
     "LICENSE"
@@ -29,7 +30,9 @@
   "types": "dist/index.d.ts",
   "scripts": {
     "build": "tsc -b tsconfig.build.json",
+    "postbuild": "downlevel-dts dist _ts3.4/dist",
     "clean": "tsc -b tsconfig.build.json --clean",
+    "postclean": "rimraf dist && rimraf _ts3.4",
     "format": "prettier --write \"./**/*.{ts,js,json,md}\" --ignore-path ../../.prettierignore",
     "lint": "eslint . --ext .js,.ts --ignore-path='../../.eslintignore'",
     "test": "jest --coverage",
@@ -50,5 +53,12 @@
   "funding": {
     "type": "opencollective",
     "url": "https://opencollective.com/typescript-eslint"
+  },
+  "typesVersions": {
+    "<3.8": {
+      "*": [
+        "ts3.4/*"
+      ]
+    }
   }
 }

--- a/packages/parser/package.json
+++ b/packages/parser/package.json
@@ -33,6 +33,7 @@
   "scripts": {
     "build": "tsc -b tsconfig.build.json",
     "clean": "tsc -b tsconfig.build.json --clean",
+    "postclean": "rimraf dist",
     "format": "prettier --write \"./**/*.{ts,js,json,md}\" --ignore-path ../../.prettierignore",
     "lint": "eslint . --ext .js,.ts --ignore-path='../../.eslintignore'",
     "test": "jest --coverage",

--- a/packages/shared-fixtures/package.json
+++ b/packages/shared-fixtures/package.json
@@ -5,6 +5,7 @@
   "scripts": {
     "build": "tsc -b tsconfig.build.json",
     "clean": "tsc -b tsconfig.build.json --clean",
+    "postclean": "rimraf dist",
     "typecheck": "tsc -p tsconfig.json --noEmit"
   }
 }

--- a/packages/typescript-estree/package.json
+++ b/packages/typescript-estree/package.json
@@ -33,6 +33,7 @@
   "scripts": {
     "build": "tsc -b tsconfig.build.json",
     "clean": "tsc -b tsconfig.build.json --clean",
+    "postclean": "rimraf dist",
     "format": "prettier --write \"./**/*.{ts,js,json,md}\" --ignore-path ../../.prettierignore",
     "lint": "eslint . --ext .js,.ts --ignore-path='../../.eslintignore'",
     "test": "jest --coverage",

